### PR TITLE
Laboratory work #4, Sofya Kuzmichyova - 21FPL2

### PR DESCRIPTION
**Pull request by sophiek19**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -4,6 +4,25 @@
 """
 from pathlib import Path
 from typing import Optional, Union
+import random
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:7]: revert to main

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py [hunk 2]
```diff
@@ -2,15 +2,31 @@
...
 
-from lab_3_keywords_textrank.main import TextEncoder, \
-    TextPreprocessor
+from lab_3_keywords_textrank.main import TextEncoder, TextPreprocessor
 
 PreprocessedSentence = tuple[str, ...]
 EncodedSentence = tuple[int, ...]
 
 
+def raise_error(obj: Any, type_name: Any, el_type: Any):
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:15]: the name is bad.

File: lab_4_summarization_textrank/main.py [hunk 3]
```diff
@@ -84,31 +108,50 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
         :return: a sequence of sentences
         """
-        pass
+        raise_error(text, str, str)
+        lst = re.findall('[.!?]\s*', text)
+        sentences = []
+        for i in range(len(lst)):
+            sentences.append(text[:text.index(lst[i]) + 1].strip())
+            if i != len(lst) - 1:
+                text = text[text.index(lst[i]) + 1::]
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:129]: Why do you use the second `:`?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py [hunk 4]
```diff
@@ -2,15 +2,31 @@
...
 
+def raise_error(obj: Any, type_name: Any, el_type: Any):
+    """
+    Checks if arguments have correct type. If object has a wrong type, raises ValueError.
+    Parameters:
+    obj: object is checked
+    type_name: object's supposed type
+    el_type: type of object's elements
+    """
+    if not (isinstance(obj, type_name) and obj is not True):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:23]: open brackets, True?

File: lab_4_summarization_textrank/main.py [hunk 5]
```diff
@@ -2,15 +2,31 @@
...
+        raise ValueError
+    if isinstance(obj, Sequence) and len(obj) != 0:
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:25]: eliminate Sequence

File: lab_4_summarization_textrank/main.py [hunk 6]
```diff
@@ -120,23 +163,36 @@ def __init__(self) -> None:
...
     def _learn_indices(self, tokens: tuple[str, ...]) -> None:
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        raise_error(tokens, tuple, str)
+        words = tuple(token for token in tokens if token not in self._word2id)
+        numbers = tuple(set(random.sample(range(1000, 10000), len(words))))
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:179]: monotonically increase from 1000 to ...

File: lab_4_summarization_textrank/main.py [hunk 7]
```diff
@@ -120,23 +163,36 @@ def __init__(self) -> None:
         """
         Constructs all the necessary attributes
         """
-        pass
+        super().__init__()
+        self._word2id = {}
+        self._id2word = {}
+        self._last = 0
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:169]: 1000

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**